go/upgrade/upgrade.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	// In case the previous startup was e.g. interruptd during the second part of the
+	// upgrade, we need to make sure that we're the same version as the previous run.
+	if u.pending.RunningVersion != "" && u.pending.RunningVersion != thisVersion {
+		return api.ErrInvalidResumingVersion
+	}
+
+	// Everything checks out, fill in the blanks.
+	u.pending.RunningVersion = thisVersion
+	_ = u.flushDescriptor()
+	u.logger.Info("loaded pending upgrade metadata",
+		"name", u.pending.Descriptor.Name,
+		"last_stage", u.pending.LastCompletedStage,
+	)
+	return nil
+}
+
+func (u *upgradeManager) flushDescriptor() error {
+	if u.pending.IsCompleted() {
+		u.logger.Info("upgrade completed, removing state",
+			"name", u.pending.Descriptor.Name,
+		)
+		err := u.store.Delete(metadataStoreKey)
+		if err == nil {
+			u.pending = nil
+		}
+		return err
+	}
+	return u.store.PutCBOR(metadataStoreKey, &u.pending)
+}
+
+func (u *upgradeManager) StartupUpgrade() error {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	if u.pending == nil || !u.pending.UpgradeEpochReached {
+		return nil
+	}
+	if !u.pending.HasStage(api.UpgradeStageStartup) {
+		// Make sure we're in order (pushing will panic otherwise).
+		u.pending.PushStage(api.UpgradeStageStartup)
+
+		u.logger.Warn("performing startup upgrade",
+			"name", u.pending.Descriptor.Name,
+			"submitted_by", u.pending.SubmittingVersion,
+			"version", u.pending.RunningVersion,
+			logging.LogEvent, api.LogEventStartupUpgrade,
+		)
+		err := u.handler.StartupUpgrade(u.ctx)
+		if err == nil {
+			// Save the updated descriptor state.
+			err = u.flushDescriptor()
+		}
+		return err
+	}
+	u.logger.Warn("startup upgrade already performed, skipping",
+		"name", u.pending.Descriptor.Name,
+		"submitted_by", u.pending.SubmittingVersion,
+		"version", u.pending.RunningVersion,
+	)
+	return nil
+}
+
+func (u *upgradeManager) ConsensusUpgrade(privateCtx interface{}, currentEpoch epochtime.EpochTime) error {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	if u.pending == nil {
+		return nil
+	}
+
+	// If we haven't reached the upgrade epoch yet, we run normally;
+	// startup made sure we're an appropriate binary for that.
+	if !u.pending.UpgradeEpochReached {
+		if currentEpoch < u.pending.Descriptor.Epoch {
+			return nil
+		}
+		u.pending.UpgradeEpochReached = true
+		if err := u.flushDescriptor(); err != nil {
+			return err
+		}
+		return api.ErrStopForUpgrade
+	}
+
+	if !u.pending.HasStage(api.UpgradeStageConsensus) {
+		// Make sure we're in order (pushing will panic otherwise).
+		u.pending.PushStage(api.UpgradeStageConsensus)
+
+		u.logger.Warn("performing consensus upgrade",
+			"name", u.pending.Descriptor.Name,
+			"submitted_by", u.pending.SubmittingVersion,
+			"version", u.pending.RunningVersion,
+			logging.LogEvent, api.LogEventConsensusUpgrade,
+		)
+		err := u.handler.ConsensusUpgrade(u.ctx, privateCtx)
+		if err == nil {
+			// Save the updated descriptor state.
+			err = u.flushDescriptor()
```

Comment:
There is one tiny problem -- assume the upgrade succeeded, and flush returned without errors. Now if the node crashes before the upgrade block is complete, the block will not be committed but the "upgrade has been done" will be committed and so will not run on restart, causing a state corruption.

I think a safer approach would be to consider the current height as an indicator of whether the upgrade has been performed or not. The consensus upgrade must be executed exactly once at the block height after the epoch transition. If we are past that height, then the upgrade has been completed, but we may re-enter that same height multiple times in case we crash at an unlucky time.